src/merkle_tree/mod.rs

```diff
@@ -1,253 +1,399 @@
 
 pub trait Config {
-    const HEIGHT: usize;
-    type H: FixedLengthCRH;
+    type LeafHash: CRH;
+    type TwoToOneHash: TwoToOneCRH;
 }
 
-/// Stores the hashes of a particular path (in order) from leaf to root.
-/// Our path `is_left_child()` if the boolean in `path` is true.
-#[derive(Derivative)]
-#[derivative(
-    Clone(bound = "P: Config"),
-    Debug(bound = "P: Config, <P::H as FixedLengthCRH>::Output: fmt::Debug")
-)]
+pub type TwoToOneDigest<P> = <<P as Config>::TwoToOneHash as TwoToOneCRH>::Output;
+pub type LeafDigest<P> = <<P as Config>::LeafHash as CRH>::Output;
+pub type TwoToOneParam<P> = <<P as Config>::TwoToOneHash as TwoToOneCRH>::Parameters;
+pub type LeafParam<P> = <<P as Config>::LeafHash as CRH>::Parameters;
+
+/// Stores the hashes of a particular path (in order) from root to leaf.
+/// For example:
+/// ```tree_diagram
+///         [A]
+///        /   \
+///      [B]    C
+///     / \   /  \
+///    D [E] F    H
+///   .. / \ ....
+///    [I] J
+/// ```
+///  Suppose we want to prove I, then `leaf_sibling_hash` is J, `auth_path` is `[C,D]`
 pub struct Path<P: Config> {
-    pub(crate) path: Vec<(Digest<P>, Digest<P>)>,
+    pub(crate) leaf_sibling_hash: LeafDigest<P>,
+    /// The sibling of path node ordered from higher layer to lower layer (does not include root node).
+    pub(crate) auth_path: Vec<TwoToOneDigest<P>>,
+    /// stores the leaf index of the node
+    pub(crate) leaf_index: usize,
 }
 
-pub type Parameters<P> = <<P as Config>::H as FixedLengthCRH>::Parameters;
-pub type Digest<P> = <<P as Config>::H as FixedLengthCRH>::Output;
-
-impl<P: Config> Default for Path<P> {
-    fn default() -> Self {
-        let mut path = Vec::with_capacity(P::HEIGHT as usize);
-        for _i in 1..P::HEIGHT as usize {
-            path.push((
-                <P::H as FixedLengthCRH>::Output::default(),
-                <P::H as FixedLengthCRH>::Output::default(),
-            ));
-        }
-        Self { path }
+impl<P: Config> Path<P> {
+    /// The position of on_path node in `leaf_and_sibling_hash` and `non_leaf_and_sibling_hash_path`.
+    /// `position[i]` is 0 (false) iff `i`th on-path node from top to bottom is on the left.
+    ///
+    /// This function simply converts `self.leaf_index` to boolean array in big endian form.
+    fn position_list(&self) -> Vec<bool> {
+        let position: Vec<_> = (0..self.auth_path.len() + 1)
+            .map(|i| ((self.leaf_index >> i) & 1) != 0)
+            .rev()
+            .collect();
+        position
     }
 }
 
 impl<P: Config> Path<P> {
+    /// Verify that a leaf is at `self.index` of the merkle tree.
+    /// * `leaf_size`: leaf size in number of bytes
+    ///
+    /// `verify` infers the tree height by setting `tree_height = self.auth_path.len() + 2`
     pub fn verify<L: ToBytes>(
         &self,
-        parameters: &<P::H as FixedLengthCRH>::Parameters,
-        root_hash: &<P::H as FixedLengthCRH>::Output,
+        leaf_hash_parameters: &LeafParam<P>,
```

Comment:
Should these parameters instead go into the Path struct?

Reply:
If those parameters typically large or small? When I wrote this, I thought if putting those parameters to path requires more memory when the verifier queries a lot of paths

Reply:
cc @Pratyush if you have thoughts on this. I'm not sure how big the parameters normally are

Reply:
Lets file an issue for this, and come back to it later?